pkg/duck/sinks.go

```diff
@@ -17,50 +17,77 @@ limitations under the License.
 package duck
 
 import (
-	"context"
 	"fmt"
 
 	corev1 "k8s.io/api/core/v1"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/client-go/dynamic"
+	"k8s.io/apimachinery/pkg/api/meta"
 
-	duckapis "github.com/knative/pkg/apis"
-	"github.com/knative/pkg/apis/duck"
+	pkgapisduck "github.com/knative/pkg/apis/duck"
 	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+
+	"github.com/knative/eventing/pkg/reconciler"
+	"github.com/knative/pkg/tracker"
 )
 
-// GetSinkURI retrieves the sink URI from the object referenced by the given
-// ObjectReference.
-func GetSinkURI(ctx context.Context, dynamicClient dynamic.Interface, sink *corev1.ObjectReference, namespace string) (string, error) {
-	if sink == nil {
+// SinkReconciler is a helper for Sources. It triggers
+// reconciliation on creation of the source's sink.
+type SinkReconciler struct {
+	tracker             tracker.Interface
+	sinkInformerFactory pkgapisduck.InformerFactory
+}
+
+// NewSinkReconciler creates and initializes a new SinkReconciler
+func NewSinkReconciler(opt reconciler.Options, callback func(string)) *SinkReconciler {
+	ret := &SinkReconciler{}
+
+	ret.tracker = tracker.New(callback, opt.GetTrackerLease())
+	ret.sinkInformerFactory = &pkgapisduck.CachedInformerFactory{
+		Delegate: &pkgapisduck.EnqueueInformerFactory{
+			Delegate: &pkgapisduck.TypedInformerFactory{
+				Client:       opt.DynamicClientSet,
+				Type:         &duckv1alpha1.AddressableType{},
+				ResyncPeriod: opt.ResyncPeriod,
+				StopChannel:  opt.StopChannel,
+			},
+			EventHandler: reconciler.Handler(ret.tracker.OnChanged),
+		},
+	}
+
+	return ret
+}
+
+// GetSinkURI registers the given object reference with the tracker and if possible,
+// retrieves the sink URI
+func (r *SinkReconciler) GetSinkURI(sinkObjRef *corev1.ObjectReference, source interface{}, sourceName string) (string, error) {
```

Comment:
Does this need a special case for K8s Services?

Reply:
@Harwayne could you explain why?

Reply:
Within the core eventing pieces (e.g. `Trigger` and `Subscription`), the subscriber is allowed to be either an Addressable or, as a special case, a K8s Service. It seems like this should be true of eventing sources as well.

Reply:
Thanks! Let me look into that.

Reply:
Added a special case for k8s Services, commit d8dcd50f6a4e4b39eaf30eaa359f47d6b0600c15
Patterned after https://github.com/knative/eventing/blob/759d58bcdd5fd672bd5b809580aae25d941eee34/pkg/duck/subscriber.go#L92